rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,221 @@
+# Leaderless DNS Load Balancing
+
+- Feature Name: Leaderless DNS Load Balancing
+- Start Date: 2024-01-17
+- RFC PR: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/pull/55)
+- Issue tracking: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Enable the MGC to manage DNS for one or more hostnames shared across one or more clusters, where a central "hub" cluster is not required or may not be available, and increase the robustness of the DNS Health checks by executing them from all the available clusters.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the MGC has limited functionality for a single cluster, or for multiple clusters without OCM. Having the MGC enabled to manage DNS in these situations significantly eases onboarding users to the advantages of the MGC.
+
+# Diagrams
+[diagrams]: #diagrams
+
+- [Overview](./0008-leaderless-dns-assets/leaderless-dns-usecases.jpg)
+- [P2P Health Checks](./0008-leaderless-dns-assets/leaderless-dns-healthchecks.jpg)
+- [DNS Pruning](./0008-leaderless-dns-assets/leaderless-dns-pruning.jpg)
+- [DNS Controller](./0008-leaderless-dns-assets/leaderless-dns-dns-operator.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster
+In a single cluster scenario, create a DNS Policy that references a DNS provider. When a matching host on a gateway is found to be in use, the IP addresses of that gateway will be created as A records for that host in the Zone file following the configuration outlined in the DNS Policy.
+
+### Single Cluster Health Checks
+If health checks are defined in the DNS Policy, these will also be created and used to remove unhealthy IPs from the Zone file, should all health checks fail, then all A records will be published to avoid an NXDOMAIN response.
+
+## Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret that specifies a unique `clusterID`, they should all reference the same zone(s) in the same DNS Provider for the same host. In each cluster, when a matching host on a gateway is found to be in use, the IP addresses of that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways IP addresses to the same zone file, without destroying each other's values.
```

Comment:
```suggestion
In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret that specifies a unique `clusterID`, they should all reference the same zone(s) in the same DNS Provider for the same host. In each cluster, when a matching listener host defined in the gateway is found that matches one of the available zones in the provider the addresses from that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways addresses to the same zone file, without destroying each other's values.
```

Reply:
Is it fair to say, that the host is the key here? If you had two gateways with 4 hosts but only one of those was shared across clusters then 3 of the hosts would be acted on as a dedicated host and 1 of the hosts as a shared host?